### DISCUSSION TITLE
Read level of digital input

### DISCUSSION THREAD

**Discussion by bennyslbs** [arendst/Tasmota#21822] [Support request] [answered]
Hi,
I have some level sensors switches that switch wery rare. Sometimes they do not register the change unless I toggles the switch.
Restart does not register the correct state.
GPIO viewer shows the same starte as reported in SENSOR = {...} and POWER/Relay
I assume the problem occurs if the ESP reboots during a change of state, but i have not verified it.

Setup:
- ESP32 with "Tasmota 14.0.0 (release-tasmota32) by Theo Arends"
- Sensors: Swith between GND and GPIO-input with pull-up
- Used GPIO's: 04, 16-19, 21-23 (8 in total)
- Each input is controlling a relay output (HW pin not used)
- Relevant Config:
  - PowerOnState 3
  - SwitchDebounce 50
  - SwitchMode1-8 1
  - Input Gpios = Switch (160-167)
  - Relay outputs (unused) = Relay (224-231)

Any way to fix this without chaning the HW?

Thanks in advance

Edit:
Berry script is OK, only requirement is that the switch state is reported to MQTT with up to around 1 second delay

**Comment by sfromis** [comment]
You cannot read level, digital means on/off only. Maybe your sensors are dirty or corroded, not reliably providing good contact.

**Reply by bennyslbs** [reply]
I don't want to read analog level but level HIGH/Low instead of detecting the edge, as I think the GPIO=Switch detects.

**Reply by sfromis** [reply, marked as answer]
Well, you **are** already seeing high-low, unrelated to edge detection, including when you use the GPIO Viewer. Tasmota is not keeping any state per Switch input, but **when** and only when it changes, Tasmota can use this event to do something.

Hence it cannot be something happening during a restart, but what the ESP8266 detect when asked. Therefore, the hardware side is where to look. Maybe the contacts are not working well, or maybe your wiring is so thin/long/noisy that you need more pullup or some filtering.

**Reply by bennyslbs** [reply]
I thought the Switch inputs was interrupt driven, but can see they are polled.

I will check my switches and add extra pull up resistors.
Funny that I have both seen sensors reporting false high and low, and in both cases it fixes the problem by togging the switches.
